Fix mob always drop exp when killed

XuZhen86 opened the pull request:
This pull request tries to fix the bug that mob always drop exp when killed.

Observed behavior:
Mobs always drop exp when killed, even drown in water without player touching them.

Expected behavior:
> A mob will not drop experience unless it dies within five seconds (100 game ticks) of an attack registered as a player hit (including tamed wolves and TNT).
[Experience - Official Minecraft Wiki](https://minecraft.gamepedia.com/Experience)

Solution:
1. Add a countdown to track the 100 game ticks.
2. Enable countdown when a valid source hit the mob.
3. Set the killer when it kills the mob.
4. Add ```AnimalTamer ``` to ```GlowPlayer ```. Set the killer to the owner of the wolf when wolf kills the mob.
5. Mob only drops exp when player is in survival mode.

Possible issues:
1. The bow and arrow is still under development, the killer setting of arrow is not implemented.
2. It gives me ```java.lang.IllegalArgumentException: Compound does not contain ByteTag "CollarColor"``` whenever I place a wolf using wolf egg?
3. The check style seems to crashed it self, I had to disable it to continue the compile.

XuZhen86 committed:
Fix mob always drop exp when killed

XuZhen86 committed:
Fixed a typo

aramperes commented on src/main/java/net/glowstone/entity/GlowLivingEntity.java:
```diff
@@ -92,6 +93,21 @@
+     */
+    @Getter
+    @Setter
+    private int hitByPlayerCountdown;
```
Wouldn't it be simpler to have a field like `playerDamageTick` that is set to the current `ticksLived` when the entity is hit, then check if `ticksLived - playerDamageTick <= 100` when killed?

aramperes commented on src/main/java/net/glowstone/entity/GlowLivingEntity.java:
```diff
@@ -806,6 +829,40 @@ public void damage(double amount, Entity source, DamageCause cause) {
+        // or damaged by a tamed wolf
+        if (
+                (source instanceof GlowPlayer
+                && ((GlowPlayer)source).getGameMode() == GameMode.SURVIVAL)
```
Should adventure mode also be checked?

aramperes commented on src/main/java/net/glowstone/entity/GlowTntPrimed.java:
```diff
@@ -24,6 +24,10 @@
 
+    @Getter
+    @Setter
+    private GlowPlayer player;
```
Add Javadoc for this

XuZhen86 committed:
Small fix
1. Simplified player damage tick determine process
2. Extracted chunks of code to isPlayerHit() and determinePlayer()
3. Improved Javadoc

XuZhen86 committed:
Style fix

aramperes commented on src/main/java/net/glowstone/entity/GlowLivingEntity.java:
```diff
@@ -730,7 +739,8 @@ public void setHealth(double health) {
                     deathEvent.getDrops().addAll(data.getItems());
-                    if (data.getExperience() > 0) {
+                    // Only drop experience when hit by a player within 5 seconds (100 game ticks)
+                    if (ticksLived - playerDamageTick <= 100 && data.getExperience() > 0) {
```
You'll probably want to have `playerDamageTick` default to -1 (in order to know if the entity wasn't hit yet), and add a check here for it being positive.

```java
if (playerDamageTick >= 0 && ticksLived - playerDamageTick <= 100 && data.getExperience() > 0)
```

XuZhen86 replied:
I made it ```private int playerDamageTick = -101;```, it solves the problem without extra code.

aramperes commented on src/main/java/net/glowstone/block/blocktype/BlockTnt.java:
```diff
@@ -14,20 +14,33 @@
 
     /**
-     * Convert a TNT block into a primed TNT entity.
+     * Convert a TNT block into a primed TNT entity with the player who ignited the tnt.
```
`tnt` -> `TNT` (in the `@param` as well)

aramperes commented on src/main/java/net/glowstone/entity/GlowLivingEntity.java:
```diff
@@ -806,6 +816,15 @@ public void damage(double amount, Entity source, DamageCause cause) {
+            playerDamageTick = ticksLived;
+        }
+
+        if (health - amount <= 0) {
```
Should this if-statement be inside of the `isPlayerHit` if-block above?

XuZhen86 replied:
Strictly saying it doesn't matter, they both determine the source. But put it inside does make more sense.

Pr0methean replied:
Put it inside. It'll save a few instructions when `!isPlayerHit(source)`

aramperes commented on src/main/java/net/glowstone/entity/GlowLivingEntity.java:
```diff
@@ -830,6 +849,47 @@ public void damage(double amount, Entity source, DamageCause cause) {
 
+    /**
+     * Checks if the incoming source of damage is a "Player Hit".
+     * @param source The incoming source of damage
```
```java
/**
 * Checks if the source of damage was caused by a player.
 *
 * @param source The source of damage
 * @return true if the source of damage was caused by a player, false otherwise.
 */
```

aramperes commented on src/main/java/net/glowstone/entity/GlowLivingEntity.java:
```diff
@@ -830,6 +849,47 @@ public void damage(double amount, Entity source, DamageCause cause) {
+        // If damaged by a TNT ignited by a player
+        } else if (source instanceof GlowTntPrimed) {
+            return
+                ((GlowTntPrimed)source).getPlayer() != null
```
Casts should have a space after the closing parentheses. There's a couple to fix here.

Ex.: `((GlowTntPrimed) source).getPlayer() != null`

Also you can store the GlowTntPrimed instance before returning because you are casting it multiple times:

```java
GlowTntPrimed tnt = (GlowTntPrimed) source;
return tnt.getPlayer() != null [...]
```

aramperes commented on src/main/java/net/glowstone/entity/GlowLivingEntity.java:
```diff
@@ -830,6 +849,47 @@ public void damage(double amount, Entity source, DamageCause cause) {
+            return (Player)((GlowTntPrimed)source).getPlayer();
+        }
+        // If been killed by a player
+        else if (source instanceof GlowPlayer) {
```
Since you are returning inside the if-blocks, the `else`s are redundant and can be omitted.

Pr0methean replied:
Should we maybe use a VicariousKillerEntity or MeansOfKilling interface to simplify all of this?

XuZhen86 replied:
I don't think that's necessary. There are only several cases that require checking vicarious killing.

aramperes commented on src/main/java/net/glowstone/entity/GlowTntPrimed.java:
```diff
@@ -24,6 +24,14 @@
+     */
+    @Getter
+    @Setter
+    private GlowPlayer player;
```
Why are this variable and getter/setter necessary? The `getSource()` method should return the player who ignited the TNT (if any), according to docs.

Pr0methean replied:
IOW:
```
public GlowPlayer getPlayer() {
    Entity source = getSource();
    return (source instanceof GlowPlayer) ? (GlowPlayer) source : null;
}
public void setPlayer(Player player) {
    source = player;
}
```

aramperes commented on src/main/java/net/glowstone/entity/GlowLivingEntity.java:
```diff
@@ -830,6 +849,47 @@ public void damage(double amount, Entity source, DamageCause cause) {
+        if (source instanceof GlowPlayer) {
+            return true;
+        // If damaged by a TNT ignited by a player
+        } else if (source instanceof GlowTntPrimed) {
```
The `else`s are redundant and can be omitted (return statement inside of the if-blocks).

aramperes commented:
The exception caused by spawning a Wolf using an egg is now fixed in 12df4a7.

XuZhen86 committed:
Merge branch 'dev' of https://github.com/GlowstoneMC/Glowstone into FixMobAlwaysDropExpWhenKilled

Pr0methean commented on src/main/java/net/glowstone/entity/GlowTntPrimed.java:
```diff
@@ -98,7 +106,11 @@ private void explode() {
+            return null;
+        } else {
+            return source;
+        }
```
`return (source != null && source.isValid()) ? source : null;`

XuZhen86 committed:
Small changes

XuZhen86 committed:
Improved Javadoc

aramperes commented on src/main/java/net/glowstone/entity/GlowLivingEntity.java:
```diff
@@ -830,6 +849,64 @@ public void damage(double amount, Entity source, DamageCause cause) {
+    }
+
+
+
```
Remove whitespace (keep only 1 line between methods)

aramperes commented on src/main/java/net/glowstone/entity/GlowPlayer.java:
```diff
@@ -195,7 +196,7 @@
  */
 @DelegateDeserialization(GlowOfflinePlayer.class)
-public class GlowPlayer extends GlowHumanEntity implements Player {
+public class GlowPlayer extends GlowHumanEntity implements Player, AnimalTamer {
```
I think this is redundant because `GlowPlayer` extends `GlowHumanEntity`, which implements `HumanEntity`, which extends `AnimalTamer`. (alternatively, `GlowPlayer -> Player -> HumanEntity -> AnimalTamer`)

XuZhen86 committed:
Small Fixes

aramperes commented on src/main/java/net/glowstone/entity/GlowTntPrimed.java:
```diff
@@ -22,8 +23,28 @@
+     *
+     * @param player Player that ignited the TNT
+     */
+    public void setPlayer(Player player) {
```
Can you remove this setter and just use `setSource` with the player? I don't think having a separate method here is necessary.

XuZhen86 committed:
Pull from dev

XuZhen86 committed:
Small Fix

XuZhen86 committed:
Fix casting from Player to GlowPlayer in GlowArrow

XuZhen86 committed:
Merge branch 'dev' of github.com:XuZhen86/Glowstone into FixMobAlwaysDropExpWhenKilled

XuZhen86 committed:
Merge branch 'dev' of https://github.com/GlowstoneMC/Glowstone into FixMobAlwaysDropExpWhenKilled

Pr0methean commented on src/main/java/net/glowstone/block/blocktype/BlockTnt.java:
```diff
@@ -7,28 +7,40 @@
      */
-    public static void igniteBlock(Block tntBlock, boolean ignitedByExplosion) {
+    public static void igniteBlock(
+        GlowBlock tntBlock, boolean ignitedByExplosion, GlowPlayer player) {
```
Why do you have to narrow tntBlock from Block to GlowBlock? Things like that make unit testing more complicated.

XuZhen86 replied:
It's more consistent within the class. All other methods use Glowblock.

mastercoms replied:
I agree with Pr0methean, it should be kept like `Block`, unless there is a specific reason for doing it otherwise. The other usages of `GlowBlock` as a method parameter are not reasons for this one to use it.

XuZhen86 committed:
Merge branch 'dev' of https://github.com/GlowstoneMC/Glowstone into FixMobAlwaysDropExpWhenKilled

XuZhen86 committed:
Small fix on parameters
